File: c++/src/agents.cc
Agents Agents::instance;

Agents &Agents::getInstance() {
	return instance;
}

void Agents::init() {
	best_percentage = 100;
}

int Agents::get_best_percentage() {
	return best_percentage;
}

void Agents::new_percentage(int percentage) {
	if(percentage < best_percentage) {
		best_percentage = percentage;
	}
}

void Agents::add(int index, Agent agent) {
	agents[index] = agent;
}

Agent Agents::random_agent() {
	return agents[rand() % NUM_AGENTS];
}

void Agents::update_state() {
	for(int i = 0; i < NUM_AGENTS; i++) {
		agents[i].update_state();
	}
}

void Agents::update() {
	for(int i = 0; i < NUM_AGENTS; i++) {
		agents[i].update();
	}
}

void Agents::draw() {
	for(int i = 0; i < NUM_AGENTS; i++) {
		agents[i].draw();
	}

}

File: ruby/agents.rb
class Agents
	def initialize()
		@agents = []

		@best_percentage = START_BEST_PERCENTAGE
	end

	def new_percentage(percentage)
		if percentage < @best_percentage
			@best_percentage = percentage
		end
	end

	def best_percentage()
		return @best_percentage
	end

	def add(agent)
		@agents.push(agent)
	end

	def draw
		@agents.each { |a| a.draw() }
	end

	def each
		@agents.each { |a| yield a }
	end

	def get_active
		active_agents = []
		@agents.each do |a|
			active_agents.push a
		end

		return active_agents
	end

	def random_agent
		return @agents[rand(@agents.length)]
	end

	@@instance = Agents.new

	def self.instance
		return @@instance
	end
end
